plutus-benchmark/bls-benchmarks/src/PlutusBenchmark/BLS12_381/Common.hs

```diff
@@ -0,0 +1,326 @@
+-- editorconfig-checker-disable-file
+{-# LANGUAGE DataKinds         #-}
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE TemplateHaskell   #-}
+{-# LANGUAGE ViewPatterns      #-}
+
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+
+{- | Approximations of the sort of computations involving BLS12-381 primitives
+ that one might wish to perform on the chain.  Real on-chain code will have
+ extra overhead, but these examples help to give us an idea of the sort of
+ computation that can feasibly be carried out within the validation budget
+ limits. -}
+module PlutusBenchmark.BLS12_381.Common ( UProg
+                                        , UTerm
+                                        , checkGroth16Verify_Haskell
+                                        , listOfSizedByteStrings
+                                        , mkGroth16VerifyScript
+                                        , mkHashAndAddG1Script
+                                        , mkHashAndAddG2Script
+                                        , mkPairingScript
+                                        , mkUncompressAndAddG1Script
+                                        , mkUncompressAndAddG2Script
+                                        , toAnonDeBruijnProg
+                                        )
+where
+import PlutusCore (DefaultFun, DefaultUni)
+import PlutusTx qualified as Tx
+import UntypedPlutusCore qualified as UPLC
+
+import PlutusTx.Prelude as Tx hiding (sort, (*))
+
+import Data.ByteString (ByteString)
+import Data.ByteString qualified as BS
+import Data.Word (Word8)
+import Hedgehog.Internal.Gen qualified as G
+import Hedgehog.Internal.Range qualified as R
+import System.IO.Unsafe (unsafePerformIO)
+
+import Prelude (fromIntegral)
+
+-------------------------------- PLC stuff--------------------------------
+
+type UTerm   = UPLC.Term    UPLC.NamedDeBruijn DefaultUni DefaultFun ()
+type UProg   = UPLC.Program UPLC.NamedDeBruijn DefaultUni DefaultFun ()
+type UDBProg = UPLC.Program UPLC.DeBruijn      DefaultUni DefaultFun ()
+
+compiledCodeToTerm
+    :: Tx.CompiledCodeIn DefaultUni DefaultFun a -> UTerm
+compiledCodeToTerm (Tx.getPlcNoAnn -> UPLC.Program _ _ body) = body
+
+{- | Remove the textual names from a NamedDeBruijn program -}
+toAnonDeBruijnProg :: UProg -> UDBProg
+toAnonDeBruijnProg (UPLC.Program () ver body) =
+    UPLC.Program () ver $ UPLC.termMapNames (\(UPLC.NamedDeBruijn _ ix) -> UPLC.DeBruijn ix) body
+
+-- Create a list containing n bytestrings of length l.  This could be better.
+listOfSizedByteStrings :: Integer -> Integer -> [ByteString]
+listOfSizedByteStrings n l = unsafePerformIO . G.sample $
+                             G.list (R.singleton $ fromIntegral n)
+                                  (G.bytes (R.singleton $ fromIntegral l))
```

Comment:
> What could go wrong? I'll add `NOINLINE` but I'm not sure that I understand the issue.

In this instance probably nothing, because `unsafePerformIO` is behind a lambda and the two bound variables are inside the `unsafePerformIO` call, but in the general case lots of things can go wrong, see the [docs](https://hackage.haskell.org/package/base-4.18.0.0/docs/System-IO-Unsafe.html#v:unsafePerformIO). Although `NOINLINE` doesn't help with most of those, so idk, feel free to ignore.

Reply:
Lot's of things can go wrong, but not because of inlining. `G.sample` is non-deterministic function, so by placing it behind `unsafePerformIO` you violate referential transparency and purity of `listOfSizedByteStrings` function. It either must be an IO action, or it should accept the seed as an argument.

Reply:
> Lot's of things can go wrong, but not because of inlining. `G.sample` is non-deterministic function, so by placing it behind `unsafePerformIO` you violate referential transparency and purity of `listOfSizedByteStrings` function. It either must be an IO action, or it should accept the seed as an argument.


This is for use in benchmarks that give us an idea of how expensive various combinations of operations are, so it's not disastrous if something goes wrong here.  I take your point though; we have stuff like this in other benchmarks too and there's an issue to tidy up the generators for those (PLT-1171).  It's probably better to do all of that at once, so I've added a note to fix this too.